langchain/tools/graphql/tool.py

```diff
@@ -0,0 +1,51 @@
+import json
+from typing import Optional
+
+from langchain.callbacks.manager import (
+    AsyncCallbackManagerForToolRun,
+    CallbackManagerForToolRun,
+)
+from langchain.tools.base import BaseTool
+from langchain.utilities.graphql import GraphQLAPIWrapper
+
+
+class BaseGraphQLTool(BaseTool):
+    """Base tool for querying a GraphQL API."""
+
+    graphql_wrapper: GraphQLAPIWrapper
+
+    name = "query_graphql"
+    description = """
+    Input to this tool is a detailed and correct GraphQL query, output is a result from the API.
+    If the query is not correct, an error message will be returned.
+    If an error is returned with 'Bad request' in it, rewrite the query and try again.
+    If an error is returned with 'Unauthorized' in it, do not try again, but tell the user to change their authentication.
+
+    Example Input: query {{ allUsers {{ id, name, email }} }}
+    """  # noqa: E501
+
+    class Config:
+        """Configuration for this pydantic object."""
+
+        arbitrary_types_allowed = True
+
+    def _run_query(self, query: str) -> str:
+        """Execute a GraphQL query and return the results."""
+        result = self.graphql_wrapper.run(query)
+        return result
+
+    def _run(
+        self,
+        tool_input: str,
+        run_manager: Optional[CallbackManagerForToolRun] = None,
+    ) -> str:
+        result = self._run_query(tool_input)
```

Comment:
could we just do `result = self.graphql_wrapper.run(tool_input)` and get rid of _run_query?

Reply:
👵👍